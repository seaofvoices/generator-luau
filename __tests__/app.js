import { expect, it, describe, jest, beforeEach } from '@jest/globals'
import { getFileContent, setupGenerator, usePolly } from './utils'

const projectName = 'new-project'
let spawnMock = jest.fn()
let answers = {}

const setupContext = async () =>
  await setupGenerator('app')
    .withArguments([projectName])
    .withAnswers(answers)
    .withSpawnMock(spawnMock)

beforeEach(async () => {
  spawnMock = jest.fn()
  answers = {
    name: 'Author Name',
    authorName: 'Author Name',
    authorEmail: 'author@write.com',
    githubOwner: 'owner',
    codeOfConductContact: 'contact@example.com',
    tools: [],
    website: '',
  }

  await usePolly('all-tools-requests')
})

describe.each(['npm', 'yarn'])('using %s', (packageManager) => {
  beforeEach(() => {
    answers.packageManager = packageManager
  })

  it('creates the package.json file', async () => {
    const context = await setupContext()
    expect(JSON.parse(getFileContent(context, 'package.json'))).toMatchSnapshot(
      {
        devDependencies: {
          '@jsdotlua/jest': expect.any(String),
          '@jsdotlua/jest-globals': expect.any(String),
          npmluau: expect.any(String),
        },
      }
    )
  })

  it('runs install and prepare command', async () => {
    await setupContext()
    if (packageManager === 'yarn') {
      expect(spawnMock).toHaveBeenCalledWith(
        'spawnSync',
        packageManager,
        ['set', 'version', 'stable'],
        expect.any(Object)
      )
    }
    expect(spawnMock).toHaveBeenCalledWith(
      'spawnSync',
      packageManager,
      ['install'],
      expect.any(Object)
    )
    expect(spawnMock).toHaveBeenCalledWith(
      'spawnSync',
      packageManager,
      ['run', 'prepare'],
      expect.any(Object)
    )
  })

  it('creates generic repo files', async () => {
    const context = await setupContext()
    expect(getFileContent(context, 'README.md')).not.toBe('')
    expect(getFileContent(context, 'CODE_OF_CONDUCT.md')).not.toBe('')
    expect(getFileContent(context, '.gitignore')).not.toBe('')
    expect(getFileContent(context, '.gitattributes')).not.toBe('')
    expect(getFileContent(context, '.npmignore')).not.toBe('')
  })

  describe('Roblox environment', () => {
    beforeEach(async () => {
      answers.luaEnvironment = 'roblox'
      answers.buildRobloxModel = true
    })

    it.each(['build-roblox-model.sh', '.luaurc'])(
      'creates a %s file',
      async (fileName) => {
        const context = await setupContext()
        expect(getFileContent(context, fileName)).toMatchSnapshot()
      }
    )

    describe.each(['lua', 'luau'])('using %s', (luaExtension) => {
      beforeEach(async () => {
        answers.luaExtension = luaExtension
      })

      it('creates an entry point to run the test in Roblox studio', async () => {
        const context = await setupContext()
        expect(
          getFileContent(context, `roblox-test.server.${luaExtension}`)
        ).not.toBe('')
      })
    })
  })

  describe('Lune environment', () => {
    beforeEach(async () => {
      answers.luaEnvironment = 'lune'
    })

    it.each(['.luaurc'])(
      'creates a %s file',
      async (fileName) => {
        const context = await setupContext()
        expect(getFileContent(context, fileName)).toMatchSnapshot()
      }
    )
  })
})
