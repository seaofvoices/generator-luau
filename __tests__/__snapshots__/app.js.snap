// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`using npm Lune environment creates a .luau-analyze.json file 1`] = `
"{
  "luau-lsp.require.mode": "relativeToFile",
  "luau-lsp.require.directoryAliases": {
    "@pkg": "node_modules/.luau-aliases",
    "@lune": "~/.lune/.typedefs/0.9.3/"
  }
}
"
`;

exports[`using npm Lune environment creates a .luaurc file 1`] = `
"{
  "languageMode": "strict",
  "lintErrors": true,
  "lint": {
    "*": true
  },
  "aliases": {
    "pkg": "./node_modules/.luau-aliases",
    "lune": "~/.lune/.typedefs/0.9.3/"
  }
}
"
`;

exports[`using npm Roblox environment creates a .luau-analyze.json file 1`] = `
"{
  "luau-lsp.require.mode": "relativeToFile",
  "luau-lsp.require.directoryAliases": {
    "@pkg": "node_modules/.luau-aliases"
  }
}
"
`;

exports[`using npm Roblox environment creates a .luaurc file 1`] = `
"{
  "languageMode": "strict",
  "lintErrors": true,
  "lint": {
    "*": true
  },
  "aliases": {
    "pkg": "./node_modules/.luau-aliases"
  }
}
"
`;

exports[`using npm Roblox environment creates a build-roblox-model.sh file 1`] = `
"#!/bin/sh

set -e

DARKLUA_CONFIG=$1
BUILD_OUTPUT=$2

npm install --omit dev
npm exec --yes --package npmluau -- npmluau

rm -rf temp
mkdir -p temp
cp -r src/ temp/
cp -rL node_modules/ temp/
cp "$DARKLUA_CONFIG" "temp/$DARKLUA_CONFIG"

./scripts/remove-tests.sh temp

rojo sourcemap model.project.json -o temp/sourcemap.json

cd temp

darklua process --config "$DARKLUA_CONFIG" src src
darklua process --config "$DARKLUA_CONFIG" node_modules node_modules

cd ..

cp model.project.json temp/

rm -f "$BUILD_OUTPUT"
mkdir -p $(dirname "$BUILD_OUTPUT")

rojo build temp/model.project.json -o "$BUILD_OUTPUT"

rm -rf temp
"
`;

exports[`using npm creates the package.json file 1`] = `
{
  "author": "Author Name <author@write.com>",
  "dependencies": {},
  "description": "",
  "devDependencies": {
    "@jsdotlua/jest": Any<String>,
    "@jsdotlua/jest-globals": Any<String>,
    "npmluau": Any<String>,
  },
  "homepage": "https://github.com/owner/new-project#readme",
  "keywords": [
    "luau",
  ],
  "license": "MIT",
  "main": "src/init.luau",
  "name": "new-project",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/owner/new-project.git",
  },
  "scripts": {
    "clean": "rm -rf node_modules build temp",
    "format": "stylua .",
    "lint": "sh ./scripts/analyze.sh && selene src",
    "lint:luau": "sh ./scripts/analyze.sh",
    "lint:selene": "selene src",
    "prepare": "npmluau",
    "style-check": "stylua . --check",
    "test:roblox": "sh ./scripts/roblox-test.sh",
    "verify-pack": "npm pack --dry-run",
  },
  "version": "0.1.0",
}
`;

exports[`using yarn Lune environment creates a .luau-analyze.json file 1`] = `
"{
  "luau-lsp.require.mode": "relativeToFile",
  "luau-lsp.require.directoryAliases": {
    "@pkg": "node_modules/.luau-aliases",
    "@lune": "~/.lune/.typedefs/0.9.3/"
  }
}
"
`;

exports[`using yarn Lune environment creates a .luaurc file 1`] = `
"{
  "languageMode": "strict",
  "lintErrors": true,
  "lint": {
    "*": true
  },
  "aliases": {
    "pkg": "./node_modules/.luau-aliases",
    "lune": "~/.lune/.typedefs/0.9.3/"
  }
}
"
`;

exports[`using yarn Roblox environment creates a .luau-analyze.json file 1`] = `
"{
  "luau-lsp.require.mode": "relativeToFile",
  "luau-lsp.require.directoryAliases": {
    "@pkg": "node_modules/.luau-aliases"
  }
}
"
`;

exports[`using yarn Roblox environment creates a .luaurc file 1`] = `
"{
  "languageMode": "strict",
  "lintErrors": true,
  "lint": {
    "*": true
  },
  "aliases": {
    "pkg": "./node_modules/.luau-aliases"
  }
}
"
`;

exports[`using yarn Roblox environment creates a build-roblox-model.sh file 1`] = `
"#!/bin/sh

set -e

DARKLUA_CONFIG=$1
BUILD_OUTPUT=$2

yarn workspaces focus --production
yarn dlx npmluau

rm -rf temp
mkdir -p temp
cp -r src/ temp/
cp -rL node_modules/ temp/
cp "$DARKLUA_CONFIG" "temp/$DARKLUA_CONFIG"

./scripts/remove-tests.sh temp

rojo sourcemap model.project.json -o temp/sourcemap.json

cd temp

darklua process --config "$DARKLUA_CONFIG" src src
darklua process --config "$DARKLUA_CONFIG" node_modules node_modules

cd ..

cp model.project.json temp/

rm -f "$BUILD_OUTPUT"
mkdir -p $(dirname "$BUILD_OUTPUT")

rojo build temp/model.project.json -o "$BUILD_OUTPUT"

rm -rf temp
"
`;

exports[`using yarn creates the package.json file 1`] = `
{
  "author": "Author Name <author@write.com>",
  "dependencies": {},
  "description": "",
  "devDependencies": {
    "@jsdotlua/jest": Any<String>,
    "@jsdotlua/jest-globals": Any<String>,
    "npmluau": Any<String>,
  },
  "homepage": "https://github.com/owner/new-project#readme",
  "keywords": [
    "luau",
  ],
  "license": "MIT",
  "main": "src/init.luau",
  "name": "new-project",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/owner/new-project.git",
  },
  "scripts": {
    "clean": "rm -rf node_modules build temp",
    "format": "stylua .",
    "lint": "sh ./scripts/analyze.sh && selene src",
    "lint:luau": "sh ./scripts/analyze.sh",
    "lint:selene": "selene src",
    "prepare": "npmluau",
    "style-check": "stylua . --check",
    "test:roblox": "sh ./scripts/roblox-test.sh",
    "verify-pack": "yarn pack --dry-run",
  },
  "version": "0.1.0",
}
`;
